// pluggy-sdk/dist/types/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CREDIT_CARD_ACCOUNT_OTHER_CREDIT_TYPES = exports.CREDIT_CARD_ACCOUNT_FEE_TYPES = exports.TransactionStatus = exports.TRANSACTION_TYPES = void 0;
/**
 * @typedef TransactionType
 * The direction of the transaction.
 * If DEBIT money going out of the account.
 * If CREDIT money going into the account.
 */
exports.TRANSACTION_TYPES = ['DEBIT', 'CREDIT'];
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["PENDING"] = "PENDING";
    TransactionStatus["POSTED"] = "POSTED";
})(TransactionStatus || (exports.TransactionStatus = TransactionStatus = {}));
exports.CREDIT_CARD_ACCOUNT_FEE_TYPES = [
    'ANNUAL_FEE',
    'ATM_WITHDRAWAL_DOMESTIC',
    'ATM_WITHDRAWAL_INTERNATIONAL',
    'EMERGENCY_CREDIT_EVALUATION',
    'CARD_REISSUE',
    'BILL_PAYMENT_FEE',
    'SMS',
    'OTHER',
];
exports.CREDIT_CARD_ACCOUNT_OTHER_CREDIT_TYPES = [
    'REVOLVING_CREDIT',
    'BILL_INSTALLMENT',
    'LOAN',
    'OTHER',
];
